File: src/main/java/com/syfe/personalfinance/dto/SavingsGoalDto.java
package com.syfe.personalfinance.dto;

import jakarta.validation.constraints.Future;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Positive;
import lombok.*;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.List;

public class SavingsGoalDto {

    @Getter
    @Setter
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class CreateGoalRequest {
        @NotBlank(message = "Goal name is required")
        private String goalName;

        @NotNull(message = "Target amount is required")
        @Positive(message = "Target amount must be a positive number")
        private BigDecimal targetAmount;

        private LocalDate startDate;

        @NotNull(message = "Target date is required")
        @Future(message = "Target date must be in the future")
        private LocalDate targetDate;
    }

    @Getter
    @Setter
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class UpdateGoalRequest {
        private String goalName;

        @Positive(message = "Target amount must be a positive number")
        private BigDecimal targetAmount;

        private LocalDate startDate;

        @Future(message = "Target date must be in the future")
        private LocalDate targetDate;
    }

    @Getter
    @Setter
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class GoalProgressResponse {
        private Long id;
        private String goalName;
        private BigDecimal targetAmount;
        private LocalDate startDate;
        private LocalDate targetDate;
        private BigDecimal currentProgress;
        private Double progressPercentage;
        private BigDecimal remainingAmount;
    }

    @Getter
    @Setter
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class GoalListResponse {
        private List<GoalProgressResponse> goals;
    }
}

File: src/main/java/com/syfe/personalfinance/dto/SavingsGoalDto.kt
package com.syfe.personalfinance.dto

import jakarta.validation.constraints.Future
import jakarta.validation.constraints.NotBlank
import jakarta.validation.constraints.NotNull
import jakarta.validation.constraints.Positive
import java.math.BigDecimal
import java.time.LocalDate

class SavingsGoalDto {

    data class CreateGoalRequest(
        @field:NotBlank(message = "Goal name is required")
        val goalName: String = "",

        @field:NotNull(message = "Target amount is required")
        @field:Positive(message = "Target amount must be a positive number")
        val targetAmount: BigDecimal? = null,

        val startDate: LocalDate? = null,

        @field:NotNull(message = "Target date is required")
        @field:Future(message = "Target date must be in the future")
        val targetDate: LocalDate? = null
    )

    data class UpdateGoalRequest(
        val goalName: String? = null,

        @field:Positive(message = "Target amount must be a positive number")
        val targetAmount: BigDecimal? = null,

        val startDate: LocalDate? = null,

        @field:Future(message = "Target date must be in the future")
        val targetDate: LocalDate? = null
    )

    data class GoalProgressResponse(
        val id: Long? = null,
        val goalName: String? = null,
        val targetAmount: BigDecimal? = null,
        val startDate: LocalDate? = null,
        val targetDate: LocalDate? = null,
        val currentProgress: BigDecimal? = null,
        val progressPercentage: Double? = null,
        val remainingAmount: BigDecimal? = null
    )

    data class GoalListResponse(
        val goals: List<GoalProgressResponse> = emptyList()
    )
}
